text_extensions_for_pandas/array/arrow_conversion.py

```diff
@@ -0,0 +1,327 @@
+        ]
+
+        pa.PyExtensionType.__init__(self, pa.struct(fields))
+
+    def __reduce__(self):
+        index_dtype = self.storage_type[self.BEGINS_NAME].type
+        metadata = self.storage_type[ArrowCharSpanType.BEGINS_NAME].metadata
+        return ArrowTokenSpanType, (index_dtype, metadata)
+
+
+def char_span_to_arrow(char_span):
+    """
+    Convert a CharSpanArray to a pyarrow.ExtensionArray with a type
+    of ArrowCharSpanType and struct as the storage type. The resulting
+    extension array can be serialized and transferred with standard
+    Arrow protocols.
+
+    :param char_span: A CharSpanArray to be converted
+    :return: pyarrow.ExtensionArray containing CharSpan data
+    """
+    # Create array for begins, ends
+    begins_array = pa.array(char_span.begin)
+    ends_array = pa.array(char_span.end)
+
+    typ = ArrowCharSpanType(begins_array.type, char_span.target_text)
+    fields = list(typ.storage_type)
+
+    storage = pa.StructArray.from_arrays([begins_array, ends_array], fields=fields)
+
+    return pa.ExtensionArray.from_storage(typ, storage)
+
+
+def arrow_to_char_span(extension_array):
+    """
+    Convert a pyarrow.ExtensionArray with type ArrowCharSpanType to
+    a CharSpanArray.
+
+    :param extension_array: pyarrow.ExtensionArray with type ArrowCharSpanType
+    :return: CharSpanArray
+    """
+    if isinstance(extension_array, pa.ChunkedArray):
+        if extension_array.num_chunks > 1:
+            raise ValueError("Only pyarrow.Array with a single chunk is supported")
+        extension_array = extension_array.chunk(0)
+
+    assert pa.types.is_struct(extension_array.storage.type)
+
+    # Get target text from the begins field metadata and decode string
+    metadata = extension_array.storage.type[ArrowCharSpanType.BEGINS_NAME].metadata
+    target_text = metadata[ArrowCharSpanType.TARGET_TEXT_KEY]
+    if isinstance(target_text, bytes):
+        target_text = target_text.decode()
+
+    # Get the begins/ends pyarrow arrays
+    begins_array = extension_array.storage.field(ArrowCharSpanType.BEGINS_NAME)
+    ends_array = extension_array.storage.field(ArrowCharSpanType.ENDS_NAME)
+
+    # Zero-copy convert arrays to numpy
+    begins = begins_array.to_numpy()
+    ends = ends_array.to_numpy()
+
+    return CharSpanArray(target_text, begins, ends)
+
+
+def token_span_to_arrow(token_span):
+    """
+    Convert a TokenSpanArray to a pyarrow.ExtensionArray with a type
+    of ArrowTokenSpanType and struct as the storage type. The resulting
+    extension array can be serialized and transferred with standard
+    Arrow protocols.
+
+    :param token_span: A TokenSpanArray to be converted
+    :return: pyarrow.ExtensionArray containing TokenSpan data
+    """
+    # Create array for begins, ends
+    token_begins_array = pa.array(token_span.begin_token)
+    token_ends_array = pa.array(token_span.end_token)
+    char_begins_array = pa.array(token_span.begin)
```

Comment:
This line of code will store the character-based begin offset of each *span* in `token_span`. You should be storing the character-based begin offset of each token in the underlying tokens behind `token_span`.

Reply:
Ahh ok, got it